Compile failed, problem with std::min

Answering these questions before submitting your bug report will help us give
you a quicker answer.  Thank you!

If one or more of these questions are not applicable, feel free to remove them.

### Does this issue affect the `google-cloud-cpp` project?
Yes

### What component of `google-cloud-cpp` is this related to?
- Cloud Bigtable.
- Cloud Storage.
- Cloud Firestore (commnunity supported).

### What version of `google-cloud-cpp` are you using?
v0.21.x
commit: 00adf2e303156ac20beb0bb38894d2136950b63e

### What compiler and version are you using?
gcc version 8.3.0 (Raspbian 8.3.0-6+rpi1) 

### What operating system and version are you using?
DietPi Buster

### What were you trying to do?
Build API

### What did you expect to see?
API to compile

### What did you see instead?
API did not compile, error ocured

### Anything else you would like us to know?

```
In file included from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h: In instantiation of ‘std::vector<long long unsigned int> google::cloud::storage::v1::internal::ComputeParallelFileUploadSplitPoints(uintmax_t, const std::tuple<_Tps ...>&) [with Options = {google::cloud::storage::v1::MinStreamSize&}; uintmax_t = long long unsigned int]’:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:1075:45:   required from ‘google::cloud::v0::StatusOr<std::vector<google::cloud::storage::v1::internal::ParallelUploadFileShard> > google::cloud::storage::v1::internal::CreateUploadShards(google::cloud::storage::v1::Client, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, std::__cxx11::string, Options&& ...) [with Options = {google::cloud::storage::v1::MinStreamSize}; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:610:68:   required from here
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:970:41: error: no matching function for call to ‘min(const unsigned int&, long long unsigned int)’
                               (std::min)(max_streams,
                               ~~~~~~~~~~^~~~~~~~~~~~~
                                          div_ceil(file_size, min_stream_size)));
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/bits/char_traits.h:39,
                 from /usr/include/c++/8/string:40,
                 from /usr/local/include/google/cloud/version.h:20,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/version.h:19,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/complex_option.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/signed_url_options.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/oauth2/credentials.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/client_options.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/oauth2/google_credentials.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:15:
/usr/include/c++/8/bits/stl_algobase.h:195:5: note: candidate: ‘template<class _Tp> const _Tp& std::min(const _Tp&, const _Tp&)’
     min(const _Tp& __a, const _Tp& __b)
     ^~~
/usr/include/c++/8/bits/stl_algobase.h:195:5: note:   template argument deduction/substitution failed:
In file included from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:970:41: note:   deduced conflicting types for parameter ‘const _Tp’ (‘unsigned int’ and ‘long long unsigned int’)
                               (std::min)(max_streams,
                               ~~~~~~~~~~^~~~~~~~~~~~~
                                          div_ceil(file_size, min_stream_size)));
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/bits/char_traits.h:39,
                 from /usr/include/c++/8/string:40,
                 from /usr/local/include/google/cloud/version.h:20,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/version.h:19,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/complex_option.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/signed_url_options.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/oauth2/credentials.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/client_options.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/oauth2/google_credentials.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:15:
/usr/include/c++/8/bits/stl_algobase.h:243:5: note: candidate: ‘template<class _Tp, class _Compare> const _Tp& std::min(const _Tp&, const _Tp&, _Compare)’
     min(const _Tp& __a, const _Tp& __b, _Compare __comp)
     ^~~
/usr/include/c++/8/bits/stl_algobase.h:243:5: note:   template argument deduction/substitution failed:
In file included from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:970:41: note:   deduced conflicting types for parameter ‘const _Tp’ (‘unsigned int’ and ‘long long unsigned int’)
                               (std::min)(max_streams,
                               ~~~~~~~~~~^~~~~~~~~~~~~
                                          div_ceil(file_size, min_stream_size)));
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/algorithm:62,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/cmake-out/external/nlohmann_json/include/google/cloud/storage/internal/nlohmann_json.hpp:37,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/nljson.h:43,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/metadata_parser.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/common_metadata.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/access_control_common.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/bucket_access_control.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/bucket_metadata.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/raw_client.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/logging_client.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/client.h:19,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/usr/include/c++/8/bits/stl_algo.h:3450:5: note: candidate: ‘template<class _Tp> _Tp std::min(std::initializer_list<_Tp>)’
     min(initializer_list<_Tp> __l)
     ^~~
/usr/include/c++/8/bits/stl_algo.h:3450:5: note:   template argument deduction/substitution failed:
In file included from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:970:41: note:   mismatched types ‘std::initializer_list<_Tp>’ and ‘unsigned int’
                               (std::min)(max_streams,
                               ~~~~~~~~~~^~~~~~~~~~~~~
                                          div_ceil(file_size, min_stream_size)));
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/8/algorithm:62,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/cmake-out/external/nlohmann_json/include/google/cloud/storage/internal/nlohmann_json.hpp:37,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/nljson.h:43,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/metadata_parser.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/common_metadata.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/access_control_common.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/bucket_access_control.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/bucket_metadata.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/raw_client.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/internal/logging_client.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/client.h:19,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:18,
                 from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/usr/include/c++/8/bits/stl_algo.h:3456:5: note: candidate: ‘template<class _Tp, class _Compare> _Tp std::min(std::initializer_list<_Tp>, _Compare)’
     min(initializer_list<_Tp> __l, _Compare __comp)
     ^~~
/usr/include/c++/8/bits/stl_algo.h:3456:5: note:   template argument deduction/substitution failed:
In file included from /home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_uploads_test.cc:16:
/home/dietpi/client/lib/comm/gcp/google-cloud-cpp/google/cloud/storage/parallel_upload.h:970:41: note:   mismatched types ‘std::initializer_list<_Tp>’ and ‘unsigned int’
                               (std::min)(max_streams,
                               ~~~~~~~~~~^~~~~~~~~~~~~
                                          div_ceil(file_size, min_stream_size)));
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/ci/kokoro/windows/build-32.bat
+++ b/ci/kokoro/windows/build-32.bat
@@ -1,0 +1,36 @@
+@REM Copyright 2020 Google LLC
+@REM
+@REM Licensed under the Apache License, Version 2.0 (the "License");
+@REM you may not use this file except in compliance with the License.
+@REM You may obtain a copy of the License at
+@REM
+@REM     http://www.apache.org/licenses/LICENSE-2.0
+@REM
+@REM Unless required by applicable law or agreed to in writing, software
+@REM distributed under the License is distributed on an "AS IS" BASIS,
+@REM WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+@REM See the License for the specific language governing permissions and
+@REM limitations under the License.
+
+REM Install Bazel using Chocolatey.
+choco install --no-progress -y bazel --version 2.0.0
+
+REM Change PATH to use chocolatey's version of Bazel
+set PATH=C:\ProgramData\chocolatey\bin;%PATH%
+
+@REM capture the version for troubleshooting
+bazel version
+@REM shutdown afterwards otherwise the server locks files
+bazel shutdown
+
+REM Configure the environment to use MSVC 2019 and then switch to PowerShell.
+call "c:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Auxiliary\Build\vcvars32.bat"
+
+REM The remaining of the build script is implemented in PowerShell.
+echo %date% %time%
+cd github\google-cloud-cpp
+powershell -exec bypass ci\kokoro\windows\build.ps1
+if %errorlevel% neq 0 exit /b %errorlevel%
+
+@echo DONE "============================================="
+@echo %date% %time%

--- a/ci/kokoro/windows/build.ps1
+++ b/ci/kokoro/windows/build.ps1
@@ -61,6 +61,20 @@ if (($BuildName -eq "cmake") -or ($BuildName -eq "cmake-debug")) {
     $env:BUILD_CACHE = "gs://cloud-cpp-kokoro-results/build-cache/google-cloud-cpp/master/vcpkg/Release/x64-windows-static/cmake.zip"
     $DependencyScript = "build-cmake-dependencies.ps1"
     $BuildScript = "build-cmake.ps1"
+} elseif ($BuildName -eq "cmake-debug-x86") {
+    $env:CONFIG = "Debug"
+    $env:GENERATOR = "Ninja"
+    $env:VCPKG_TRIPLET = "x86-windows-static"
+    $env:BUILD_CACHE = "gs://cloud-cpp-kokoro-results/build-cache/google-cloud-cpp/master/vcpkg/Release/x86-windows-static/cmake-debug.zip"
+    $DependencyScript = "build-cmake-dependencies.ps1"
+    $BuildScript = "build-cmake.ps1"
+} elseif ($BuildName -eq "cmake-release-x86") {
+    $env:CONFIG = "Release"
+    $env:GENERATOR = "Ninja"
+    $env:VCPKG_TRIPLET = "x86-windows-static"
+    $env:BUILD_CACHE = "gs://cloud-cpp-kokoro-results/build-cache/google-cloud-cpp/master/vcpkg/Release/x86-windows-static/cmake-release.zip"
+    $DependencyScript = "build-cmake-dependencies.ps1"
+    $BuildScript = "build-cmake.ps1"
 } elseif ($BuildName -eq "bazel") {
     $DependencyScript = "build-bazel-dependencies.ps1"
     $BuildScript = "build-bazel.ps1"

--- a/google/cloud/storage/parallel_upload.h
+++ b/google/cloud/storage/parallel_upload.h
@@ -966,9 +966,10 @@ std::vector<std::uintmax_t> ComputeParallelFileUploadSplitPoints(
                                .value();
 
   std::size_t const wanted_num_streams =
-      (std::max<std::size_t>)(1,
-                              (std::min<std::size_t>)(max_streams,
-                                         div_ceil(file_size, min_stream_size)));
+      (std::max<std::size_t>)(1, (std::min<std::size_t>)(max_streams,
+                                                         div_ceil(
+                                                             file_size,
+                                                             min_stream_size)));
 
   std::uintmax_t const stream_size =
       (std::max<std::uintmax_t>)(1, div_ceil(file_size, wanted_num_streams));

--- a/google/cloud/storage/parallel_upload.h
+++ b/google/cloud/storage/parallel_upload.h
@@ -967,7 +967,7 @@ std::vector<std::uintmax_t> ComputeParallelFileUploadSplitPoints(
 
   std::size_t const wanted_num_streams =
       (std::max<std::size_t>)(1,
-                              (std::min)(max_streams,
+                              (std::min<std::size_t>)(max_streams,
                                          div_ceil(file_size, min_stream_size)));
 
   std::uintmax_t const stream_size =

